pkg/imgpkg/image/keychain.go

```diff
@@ -0,0 +1,208 @@
+const (
+	// ErrCredentialsNotFound standardizes the not found error, so every helper returns
+	// the same message and docker can handle it properly.
+	errCredentialsNotFoundMessage = "credentials not found in native keychain"
+	// ErrCredentialsMissingServerURL and ErrCredentialsMissingUsername standardize
+	// invalid credentials or credentials management operations
+	errCredentialsMissingServerURLMessage = "no credentials server URL"
+	errCredentialsMissingUsernameMessage  = "no credentials username"
+
+	globalPrefix = "IMGPKG_REGISTRY"
+)
+
+func Keychain(username string, password string, token string, anon bool) regauthn.Keychain {
+	return regauthn.NewMultiKeychain(customRegistryKeychain{username, password, token, anon},
+		&envKeychain{globalPrefix: globalPrefix})
+}
+
+type customRegistryKeychain struct {
+	Username string
+	Password string
+	Token    string
+	Anon     bool
+}
+
+var _ regauthn.Keychain = customRegistryKeychain{}
+
+func (k customRegistryKeychain) Resolve(res regauthn.Resource) (regauthn.Authenticator, error) {
+	switch {
+	case len(k.Username) > 0:
+		return &regauthn.Basic{Username: k.Username, Password: k.Password}, nil
+	case len(k.Token) > 0:
+		return &regauthn.Bearer{Token: k.Token}, nil
+	case k.Anon:
+		return regauthn.Anonymous, nil
+	default:
+		return retryDefaultKeychain(func() (regauthn.Authenticator, error) {
+			return regauthn.DefaultKeychain.Resolve(res)
+		})
+	}
+}
+
+func retryDefaultKeychain(doFunc func() (regauthn.Authenticator, error)) (regauthn.Authenticator, error) {
+	var auth regauthn.Authenticator
+	var lastErr error
+
+	for i := 0; i < 5; i++ {
+		auth, lastErr = doFunc()
+		if lastErr == nil {
+			return auth, nil
+		}
+
+		if strings.Contains(lastErr.Error(), errCredentialsNotFoundMessage) || strings.Contains(lastErr.Error(), errCredentialsMissingUsernameMessage) || strings.Contains(lastErr.Error(), errCredentialsMissingServerURLMessage) {
+			return auth, lastErr
+		}
+
+		time.Sleep(2 * time.Second)
+	}
+	return auth, fmt.Errorf("Retried 5 times: %s", lastErr)
+}
+
+type envKeychain struct {
+	globalPrefix string
+
+	infos       []envKeychainInfo
+	collectErr  error
+	collected   bool
+	collectLock sync.Mutex
+}
+
+var _ regauthn.Keychain = &envKeychain{}
+
+func (k *envKeychain) Resolve(target regauthn.Resource) (regauthn.Authenticator, error) {
+	infos, err := k.collect()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, info := range infos {
+		if info.Hostname == target.RegistryStr() {
+			return regauthn.FromConfig(regauthn.AuthConfig{
+				Username:      info.Username,
+				Password:      info.Password,
+				IdentityToken: info.IdentityToken,
+				RegistryToken: info.RegistryToken,
+			}), nil
+		}
+	}
+
+	return regauthn.Anonymous, nil
+}
+
+func (k *envKeychain) collect() ([]envKeychainInfo, error) {
+	k.collectLock.Lock()
+	defer k.collectLock.Unlock()
+
+	if k.collected || k.collectErr != nil {
+		return append([]envKeychainInfo{}, k.infos...), k.collectErr
+	}
```

Comment:
```suggestion
	if k.collected {
		return append([]envKeychainInfo{}, k.infos...), nil
	}

	if k.collectErr != nil {
		return nil, k.collectErr
	}
```
I am suggesting this change because it is clearer, at least in my head these 2 possible states. If not you will have to look through the code to see if there is a case where an error has occurred and we collected the keychain information.